Sources/NIOHPACK/StringRing.swift

```diff
@@ -0,0 +1,355 @@
+    /// This method is likely more efficient than the one operating on plain `Collection` as it will use `memcpy` to copy all the bytes in one go.
+    ///
+    /// - parameters:
+    ///     - bytes: A `ContiguousCollection` of `UInt8` to be written.
+    /// - returns: The number of bytes written or `bytes.count`.
+    @discardableResult
+    @_inlineable @_versioned
+    mutating func write<C: ContiguousCollection>(bytes: C) throws -> Int where C.Element == UInt8 {
+        let totalCount = bytes.count
+        guard totalCount <= self.writableBytes else {
+            throw RingBufferError.BufferOverrun(amount: Int(totalCount - self._storage.capacity))
+        }
+        
+        let firstPart = self._storage.capacity - self._ringTail
+        if firstPart >= bytes.count {
+            _storage.set(bytes: bytes, at: self._ringTail)
+        } else {
+            let secondPart = bytes.count - firstPart
+            bytes.withUnsafeBytes { ptr in
+                _storage.set(bytes: UnsafeRawBufferPointer(start: ptr.baseAddress!, count: firstPart), at: self._ringTail)
+                _storage.set(bytes: UnsafeRawBufferPointer(start: ptr.baseAddress!.advanced(by: firstPart), count: secondPart), at: 0)
+            }
+        }
+        
+        self.moveTail(forwardBy: totalCount)
+        return totalCount
+    }
+    
+    /// Write `bytes`, a `Collection` of `UInt8` into this `StringRing`. Moves the writer index forward by the number of bytes written.
+    ///
+    /// - parameters:
+    ///     - bytes: A `Collection` of `UInt8` to be written.
+    /// - returns: The number of bytes written or `bytes.count`.
+    @discardableResult
+    @_inlineable @_versioned
+    mutating func write<C : Collection>(bytes: C) throws -> Int where C.Element == UInt8 {
+        assert(!([Array<C.Element>.self, StaticString.self, ContiguousArray<C.Element>.self, UnsafeRawBufferPointer.self, UnsafeBufferPointer<UInt8>.self].contains(where: { (t: Any.Type) -> Bool in t == type(of: bytes) })),
+               "called the slower set<S: Collection> function even though \(C.self) is a ContiguousCollection")
+        guard bytes.count <= self.writableBytes else {
+            throw RingBufferError.BufferOverrun(amount: Int(self._storage.capacity) - bytes.underestimatedCount)
+        }
+        
+        // is this a contiguous write?
+        let contiguousWriteSpace = self._contiguousWriteSpace
+        guard bytes.count <= contiguousWriteSpace else {
+            // multiple writes -- split and recurse
+            let firstPart = bytes.prefix(contiguousWriteSpace)
+            let secondPart = bytes.dropFirst(contiguousWriteSpace)
+            return try self.write(bytes: firstPart) + self.write(bytes: secondPart)
+        }
+        
+        // single write, should be straightforward
+        self._storage.withVeryUnsafeBytes { ptr in
+            let targetAddr = UnsafeMutableRawPointer(mutating: ptr.baseAddress!.advanced(by: self._ringTail))
```

Comment:
I think this should be

```swift
let tailPtr = UnsafeMutableRawBufferPointer(rebasing: ptr[(ptr.startIndex + self._ringTail)...]) // given that we know we're not dealing with a slice here, we could drop the `ptr.startIndex +`
tailPtr.copyBytes(from: bytes)
```

Reply:
The compiler makes be do this in two steps, but yeah, I like this API more:

```swift
let mutable = UnsafeMutableRawBufferPointer(mutating: ptr)
let tailPtr = UnsafeMutableRawBufferPointer(rebasing: mutable[(mutable.startIndex + self._ringTail)...])
tailPtr.copyBytes(from: bytes)
```